### PR TITLE
Task/ug feature update

### PR DESCRIPTION
**Pull request by rhornung67**
# Summary

- This PR cleans up and updates the feature sections of the user guide to be more complementary to the recent tutorial content rework.
- This is a draft PR. There is more work to do.

The generated docs for this PR are here: https://raja.readthedocs.io/en/task-ug-feature-update/sphinx/user_guide/features.html

### CODE REVIEW

**Review by artv3 — APPROVED** [review]

File: docs/sphinx/user_guide/feature/loop_basic.rst [hunk 1]
```diff
@@ -103,15 +78,38 @@ simpler and less verbose for that use case.
...
 A ``RAJA::kernel`` template provides ways to compose and execute arbitrary 
-loop nests and other complex kernels. To introduce the RAJA *kernel* interface,
-consider a (N+1)-level C-style loop nest::
+loop nests and other complex kernels. 
+The ``RAJA::kernel`` interface employs similar concepts to ``RAJA::forall``
+but extends it to support much more complex kernel structures.
+Each ``RAJA::kernel`` method is a template that takes an *execution policy* 
+type template parameter. The execution policy can be an arbitrarily complex
+sequence of nested templates that define a kernel execution pattern.
+loop nest. In its simplest form, ``RAJA::kernel`` takes two arguments:
```
Comment (MrBurmark) [inline docs/sphinx/user_guide/feature/loop_basic.rst:106]: typo, `loop nest.`?

File: docs/sphinx/user_guide/feature/loop_basic.rst [hunk 2]
```diff
@@ -189,26 +186,27 @@ Each ``For`` statement takes three template parameters:
...
 of them, with ``RAJA::statement::Lambda`` types placed appropriately in the
 execution policy, to construct a loop kernel. For example, placing 
 ``RAJA::statement::Lambda`` types between ``RAJA::statement::For`` statements 
 enables non-perfectly nested loops.
 
-RAJA offers two types of lambda statements. The first as illustratated
-above, requires that each lambda expression passed to a ``RAJA::kernel`` method
+RAJA offers two types of ``RAJA::statement::Lambda`` statements. The simplest 
+for illustrated
+above requires that each lambda expression passed to a ``RAJA::kernel`` method
```
Comment (MrBurmark) [inline docs/sphinx/user_guide/feature/loop_basic.rst:199]: typo, spacing

File: docs/sphinx/user_guide/feature/resource.rst [hunk 3]
```diff
@@ -185,101 +242,112 @@ Finally, the resource type can be deduced from an execution policy::
...
+complete::
 
-The usage allows one to set up dependencies between resource objects and 
-``RAJA::forall`` calls.
+  my_res.wait_for(&e);
+
+All methods listed above near the beginning of the RAJA resource discussion
+return an event object so users can access the event associated with the 
+method call. This allows one to set up dependencies between resource objects 
+and operations and define and control asynchronous execution patterns.
```
Comment (mdavis36) [inline docs/sphinx/user_guide/feature/resource.rst:268]:
```suggestion
and operations, as well as define and control asynchronous execution patterns.
```

File: docs/sphinx/user_guide/feature/resource.rst [hunk 4]
```diff
@@ -185,101 +242,112 @@ Finally, the resource type can be deduced from an execution policy::
...
    :end-before: _raja_res_k3_end
    :language: C++
-    
-We can enqueue a memcpy operation on ``res_gpu1`` to move data from the device 
-to the host:
+
+Next, we enqueue a memcpy operation on the resource ``res_gpu1`` to copy 
+the GPU array ``d_array`` to the host array ``h_array``. **I think we need to
+add a wait call or a device sync call before the memcpy operation since the
+kernel execution is asynchronous --RDH**:
```
Comment (mdavis36) [inline docs/sphinx/user_guide/feature/resource.rst:343]: We don't need a wait call as the memcpy is enqueued on the same stream as the final raja kernel. If the memcpy was enqueued on `res_gpu2` we would need a wait

File: docs/sphinx/user_guide/feature/loop_basic.rst (hunk 2 above)
Comment (rhornung67) [reply docs/sphinx/user_guide/feature/loop_basic.rst:199]: fixed

File: docs/sphinx/user_guide/feature/resource.rst [hunk 5]
```diff
@@ -12,169 +12,226 @@
...
+          above, the *default* resource type associated with the execution
+          policy is used in the internal implementation.
+
+When passing a CUDA or HIP resource, the method will execute asynchronously 
+on a GPU stream. Currently, CUDA and HIP are the only resource types that 
+enable asynchronous threading. 
+
+.. note:: Support for OpenMP CPU multithreading, which would use the 
+          ``RAJA::resources::Host`` resource type, and OpenMP target offload,
+          which would use the ``RAJA::resources::Omp`` resource type, are
```
[the remark this answers is not in the document]
Comment (rhornung67) [reply docs/sphinx/user_guide/feature/resource.rst:179]: fixed

File: docs/sphinx/user_guide/feature/resource.rst (hunk 3 above)
Comment (rhornung67) [reply docs/sphinx/user_guide/feature/resource.rst:268]: changed

File: docs/sphinx/user_guide/feature/resource.rst (hunk 4 above)
Comment (rhornung67) [reply docs/sphinx/user_guide/feature/resource.rst:343]: comment removed.

**Comment by rhornung67** [comment]
@MrBurmark @artv3 @mdavis36 @rchen20 this should be good to merge after your reviews and approvals. Please note that I will have one more doc update PR coming soon to get merged before the release.

**Comment by rhornung67** [comment]
@MrBurmark @artv3 @rchen20 @mdavis36 I've addressed all the comments on this PR that you provided. Any issues or concerns with me merging it?

**Comment by MrBurmark** [comment]
I'm not concerned at all.

**Review by mdavis36 — APPROVED** [review]

**Review by MrBurmark — APPROVED** [review]

**Pull request merged**